end-to-end-tests/src/helpers/ctx.rs

```diff
@@ -84,32 +86,168 @@ pub fn nexus_addr() -> SocketAddr {
-                .internal_services_ip_pool_ranges
+    // Otherwise, use the RSS configuration to find the DNS server, silo name,
+    // and delegated DNS zone name.  Use this to look up Nexus's IP in the
+    // external DNS server.
+    //
+    // First, load the RSS configuration file.
+    let config = rss_config().unwrap();
+
+    // From config-rss.toml, grab the first address from the configured services
+    // IP pool as the DNS server's IP address.
+    let dns_ip = config
+        .internal_services_ip_pool_ranges
+        .iter()
+        .flat_map(|range| range.iter())
+        .next()
+        .ok_or_else(|| {
+            anyhow!(
+                "failed to get first IP from internal service \
+                pool in {}",
+                RSS_CONFIG_PATH,
+            )
+        })
+        .unwrap();
+    let dns_addr = SocketAddr::from((dns_ip, 53));
+
+    // Resolve the DNS name of the recovery Silo that ought to have been created
+    // already.  This could take a few seconds, since it's asynchronous with the
+    // rack initialization request.
+    let silo_name = &config.recovery_silo.silo_name;
+    let dns_name = format!(
+        "{}.sys.{}",
+        silo_name.as_str(),
+        &config.external_dns_zone_name
+    );
+
+    let mut resolver_config = ResolverConfig::new();
+    resolver_config.add_name_server(NameServerConfig {
+        socket_addr: dns_addr,
+        protocol: Protocol::Udp,
+        tls_dns_name: None,
+        trust_nx_responses: false,
+        bind_addr: None,
+    });
+
+    let resolver =
+        TokioAsyncResolver::tokio(resolver_config, ResolverOpts::default())
+            .context("failed to create resolver")
+            .unwrap();
+
+    wait_for_condition::<_, anyhow::Error, _, _>(
+        || async {
+            let addr = resolver
+                .lookup_ip(&dns_name)
+                .await
+                .map_err(|e| match e.kind() {
+                    ResolveErrorKind::NoRecordsFound { .. }
+                    | ResolveErrorKind::Timeout => CondCheckError::NotYet,
+                    _ => CondCheckError::Failed(anyhow::Error::new(e).context(
+                        format!("resolving {:?} from {}", dns_name, dns_addr),
+                    )),
+                })?
                 .iter()
-                .flat_map(|range| range.iter())
-                .skip(1)
                 .next()
-            {
-                return (addr, 80).into();
-            }
-        }
-    }
-
-    ([192, 168, 1, 21], 80).into()
+                .ok_or(CondCheckError::NotYet)?;
+            Ok(SocketAddr::from((addr, 80)))
+        },
+        &Duration::from_secs(1),
+        &Duration::from_secs(300),
+    )
+    .await
+    .unwrap_or_else(|e| panic!("failed to get Nexus addr: {}", e))
 }
 
-fn get_base_url() -> String {
+async fn get_base_url() -> String {
     // Check $OXIDE_HOST first.
     if let Ok(host) = std::env::var("OXIDE_HOST") {
         return host;
     }
```

Comment:
Remove this since `nexus_addr` seems to be checking it already?

Reply:
I think this was true before as well.  I went ahead and made this change though.